### PR TITLE
Switch cronjobs to systemd timers

### LINKED ISSUE #6705: Convert cron jobs into systemd timers
## Description

Port our 2 cron jobs into systemd timers

## Rationale

* systemd timers are nicer to deal with, they have more configuration knobs and provide more diagnostic output
* timers are defined in files, so manipulating them is much easier than needing to use something like ansible comments for crontabs
  * consider that efb2c766fe388c9cb22ac8c7bb5c89c214b7ddce would've just been adjusting the `User=www-data` line instead of needing to write and test a hacky bash migration step

### LINKED ISSUE #6748: OSSEC sending daily login reports, even if a user hasn't logged in
## Description

Ever since the 2.5.1 release, OSSEC sends a daily email showing either one or two logins on the app server. After investigation, it appears that this is happening as a result of changes made in [this commit](https://github.com/freedomofpress/securedrop/commit/411946a2f4336e1e18ac88259a9c07ae268df09f) which changes it so that OSSEC checks run as as the `www-data` user.

Because it now sends out these false alarms, the daily login report is less useful as a monitoring tool.

## Steps to Reproduce

This behavior happens automatically once per day, as long as the server is running and OSSEC alerts are set up correctly.

You can see where this happens by inspecting `/var/log/auth.log`. A line will appear, containing:

```
USER=www-data ; COMMAND=/opt/venvs/securedrop-app-code/bin/python3 /var/www/securedrop/manage.py check-disconnected-[db|fs]-submissions
```

followed by:

```
session opened for user www-data by (uid=0)
```

## Expected Behavior

A daily login report is only sent in the event of an actual login by an end-user.

### PR DESCRIPTION
**Pull request by legoktm**
## Status

Ready for review

## Description of Changes

systemd timers are easier to define and offer more flexibility in running commands. For example, 9169606decd3c81c would've been a 2 line change instead of a complex and fragile bash migration.

We clear the old crontabs in the postinst and let debhelper enable the timers.

Turns out our usage of `dh_systemd_enable` and `dh_systemd_start` was wrong, because we were calling it multiple times without specifying explicit unit files, it was adding the same stanza to the postinst/postrm multiple times. We now invoke it explicitly for each unit we want enabled/started. Hopefully when we switch to compat 12/dh_installsystemd we don't need to manually list each one.

Fixes #6705.
Fixes #6748.

## Testing

* [x] Run `make build-debs`. Copy the securedrop-app-code deb into a temporary directory (so it doesn't get messy) and extract the maintainer scripts with `ar x securedrop-app-code_2.6.0\~rc1+focal_amd64.deb && tar xvf control.tar.xz`. Verify that the 2 timers and 4 pre-existing service units all have identical stanzas in the `postinst`, `postrm`, and `prerm`.
  * [x] You can also run `make build-debs` without this PR and diffoscope against the new build with this PR, but I found enough stuff got re-arranged that the diff wasn't very useful to look at.
* [ ] Set up a staging/prod instance with either 2.5.1 or develop before this PR. Then with this PR, run `make build-debs` and install the securedrop-app-code deb with `sudo apt install ./securedrop-app-code_2.6.0~rc1+focal_amd64.deb`
* [ ] Run `systemctl list-timers`, see the 2 new timers appear
   * [ ]  `securedrop-clean-tmp` should be scheduled for 00:00 UTC.
   * [ ] `securedrop-submissions-today` should be scheduled for 1 hour before the reboot. Easiest way to verify this is add the amount of hours left (e.g. mine says 9h left) with the uptime (mine is 14h), it should add up to 23h.
* [ ] Manually create a backdated temporary file to be deleted: `sudo -u www-data touch /var/lib/securedrop/tmp/delete-me -mt 2303300000`
* [ ] Run `sudo crontab -l` and `sudo crontab -u www-data -l`, both should say "no crontab for root/www-data"
* [ ] Run `systemctl status securedrop-clean-tmp.service`, verify it is not enabled and not active. Then run  `systemctl status securedrop-clean-tmp.timer`, verify it is active and enabled.
  * [ ] Repeat for `securedrop-submissions-today.{service,timer}`
* [ ] Wait a day
* [ ] verify you receive the next "Submissions in the past 24h" email at the same time as before.
* [ ] Run `sudo -u www-data ls -l /var/lib/securedrop/tmp/` and verify the `delete-me` file is gone.
* [ ] Run `sudo cat /var/log/auth.log | grep cron | grep www-data`, verify the lines like `app CRON[268825]: pam_unix(cron:session): session opened for user www-data by (uid=0)` have stopped.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
     * The postinst will clear any cron entries and the packaging sets up the systemd services and timers
2. New installs.
    * The postinst will gracefully skip over the fact no cron entries exist, and the packaging still sets up the systemd services + timers 

## Checklist

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass
- [ ] I have written a test plan and validated it for this PR
- [x] These changes do not require documentation

### CODE REVIEW

**Comment by legoktm** [comment]
Still in draft because I need to wait a day to ensure the timers + units are executed properly and on time. I'm also not sure whether the `OnBootSec=23h` hack is acceptable. E.g. if the instance reboots in the middle of the day for whatever reason (power loss, maintenance, etc.), then that day's notification will get skipped.

A more robust approach might be to run the timer every hour and have the script check the time and do nothing if it's not correct.

File: securedrop/debian/app-code/lib/systemd/system/securedrop-submissions-today.service [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=job to clean SecureDrop tmp dir daily
```
Comment (cfm) [inline securedrop/debian/app-code/lib/systemd/system/securedrop-submissions-today.service]: Looks like copypasta from `securedrop-clean-tmp.service`.

File: molecule/testinfra/app/test_appenv.py [hunk 2]
```diff
@@ -105,16 +105,9 @@ def test_ensure_logo(host):
         assert f.group == "root"
 
 
-def test_securedrop_tmp_clean_cron(host):
-    """Ensure securedrop tmp clean cron job in place"""
-    with host.sudo():
-        cronlist = host.run("crontab -u www-data -l").stdout
-        cronjob = f"@daily {sdvars.securedrop_code}/manage.py clean-tmp"
-        assert cronjob in cronlist
```
Comment (cfm) [inline molecule/testinfra/app/test_appenv.py:113]: Would it be worthwhile to add tests for the expected systemd units, especially after the additional changes in `securedrop/debian/rules`?

File: securedrop/debian/rules [hunk 3]
```diff
@@ -63,11 +63,11 @@ override_dh_installinit:
...
+	dh_systemd_enable --no-enable securedrop-clean-tmp.service
+	dh_systemd_enable
 
 override_dh_systemd_start:
-	dh_systemd_start --name=securedrop_rqrequeue
-	dh_systemd_start --name=securedrop_rqworker
-	dh_systemd_start --name=securedrop_shredder
+	dh_systemd_start --no-start securedrop-submissions-today.service
+	dh_systemd_start --no-start securedrop-clean-tmp.service
+	dh_systemd_start
```
Comment (cfm) [inline securedrop/debian/rules]: `dh_systemd_{enable,start} --no-{enable,start}` is a bit of a riddle.  Would you mind documenting these overrides?  :-)

**Review by cfm — CHANGES_REQUESTED** [review]
I've started with a code review (comments inline) and an inspection of the `securedrop-app-code` package according to the test plan:

* [x] Run `make build-debs`. Copy the securedrop-app-code deb into a temporary directory (so it doesn't get messy) and extract the maintainer scripts with `ar x securedrop-app-code_2.6.0\~rc1+focal_amd64.deb && tar xvf control.tar.xz`. Verify that the 2 timers and 4 pre-existing service units all have identical stanzas in the `postinst`, `postrm`, and `prerm`.

Identical *within* each script, not across scripts, per <https://gist.github.com/cfm/389f7ea72a526e558f7b6b2bb9a3224e>.  :-)

Tomorrow I'll spin up a staging instance for testing.

**Comment by cfm** [comment]
I've kicked off the test plan in both staging VMs (for cleanliness) and QA hardware (for e-mail notifications) in <https://pad.riseup.net/p/securedrop-6780-cfm>.  I'll have results tomorrow.

File: securedrop/debian/rules (hunk 3 above)
Comment (legoktm) [reply securedrop/debian/rules]: Yes...

File: molecule/testinfra/app/test_appenv.py (hunk 2 above)
Comment (legoktm) [reply molecule/testinfra/app/test_appenv.py:113]: I struggle with balancing what things the package does that we can just trust that it does vs those we should assert. And I don't want to end up in the state where we have two lists of all the units we install and then test..

What would you expect/want a test to check?

Comment (cfm) [reply molecule/testinfra/app/test_appenv.py:113]: Very fair.  I can get over my knee-jerk reaction at the asymmetry of—

```
‒ cron jobs
‒ tests for cron jobs
+ systemd timers
```

—given that the latter's declarative trustworthiness is our whole motivation for this switch .  :-)

**Review by cfm — APPROVED** [review]
Trivial changes since yesterday are trivial, and the test plan checked out nicely on both staging VMs and QA hardware:

* [x] Set up a staging/prod instance with either 2.5.1 or develop before this PR. Then with this PR, run `make build-debs` and install the securedrop-app-code deb with `sudo apt install ./securedrop-app-code_2.6.0~rc1+focal_amd64.deb`

```sh-session
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop-clean-tmp.timer
+ deb-systemd-helper --quiet was-enabled securedrop-clean-tmp.timer
+ deb-systemd-helper enable securedrop-clean-tmp.timer
Created symlink /etc/systemd/system/timers.target.wants/securedrop-clean-tmp.timer → /lib/systemd/system/securedrop-clean-tmp.timer.
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop-submissions-today.timer
+ deb-systemd-helper --quiet was-enabled securedrop-submissions-today.timer
+ deb-systemd-helper enable securedrop-submissions-today.timer
Created symlink /etc/systemd/system/timers.target.wants/securedrop-submissions-today.timer → /lib/systemd/system/securedrop-submissions-today.timer.
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop_rqrequeue.service
+ deb-systemd-helper --quiet was-enabled securedrop_rqrequeue.service
+ deb-systemd-helper enable securedrop_rqrequeue.service
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop_rqworker.service
+ deb-systemd-helper --quiet was-enabled securedrop_rqworker.service
+ deb-systemd-helper enable securedrop_rqworker.service
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop_shredder.service
+ deb-systemd-helper --quiet was-enabled securedrop_shredder.service
+ deb-systemd-helper enable securedrop_shredder.service
+ '[' configure = configure ']'
+ deb-systemd-helper unmask securedrop_source_deleter.service
+ deb-systemd-helper --quiet was-enabled securedrop_source_deleter.service
+ deb-systemd-helper enable securedrop_source_deleter.service
+ '[' configure = configure ']'
+ '[' -d /run/systemd/system ']'
+ systemctl --system daemon-reload
+ deb-systemd-invoke start securedrop-clean-tmp.service securedrop-clean-tmp.timer securedrop-submissions-today.service securedrop-submissions-today.timer securedrop_rqrequeue.service securedrop_rqworker.service securedrop_shredder.service securedrop_source_deleter.service
securedrop-clean-tmp.service is a disabled or a static unit, not starting it.
securedrop-submissions-today.service is a disabled or a static unit, not starting it.
```

* [x] Run `systemctl list-timers`, see the 2 new timers appear
   * [x]  `securedrop-clean-tmp` should be scheduled for 00:00 UTC.
   * [x] `securedrop-submissions-today` should be scheduled for 1 hour before the reboot. Easiest way to verify this is add the amount of hours left (e.g. mine says 9h left) with the uptime (mine is 14h), it should add up to 23h.

```sh-session
vagrant@app-staging:~$ systemctl list-timers | grep "securedrop"
Thu 2023-05-25 00:00:00 UTC 4h 53min left n/a                         n/a       securedrop-clean-tmp.timer         securedrop-clean-tmp.service        
Thu 2023-05-25 17:51:57 UTC 22h left      n/a                         n/a       securedrop-submissions-today.timer securedrop-submissions-today.service
vagrant@app-staging:~$ uptime
 19:07:19 up 15 min,  2 users,  load average: 0.13, 0.10, 0.04
```

* [x] Manually create a backdated temporary file to be deleted: `sudo -u www-data touch /var/lib/securedrop/tmp/delete-me -mt 2303300000`
* [x] Run `sudo crontab -l` and `sudo crontab -u www-data -l`, both should say "no crontab for root/www-data"

```sh-session
vagrant@app-staging:~$ sudo crontab -l
no crontab for root
vagrant@app-staging:~$ sudo crontab -u www-data -l
no crontab for www-data
```

* [x] Run `systemctl status securedrop-clean-tmp.service`, verify it is not enabled and not active. Then run  `systemctl status securedrop-clean-tmp.timer`, verify it is active and enabled.
  * [x] Repeat for `securedrop-submissions-today.{service,timer}`

```sh-session
vagrant@app-staging:~$ systemctl status securedrop-clean-tmp.service
● securedrop-clean-tmp.service - job to clean SecureDrop tmp dir daily
     Loaded: loaded (/lib/systemd/system/securedrop-clean-tmp.service; static; vendor preset: enabled)
     Active: inactive (dead)
TriggeredBy: ● securedrop-clean-tmp.timer
vagrant@app-staging:~$ systemctl status securedrop-clean-tmp.timer
● securedrop-clean-tmp.timer - Run clean-tmp daily
     Loaded: loaded (/lib/systemd/system/securedrop-clean-tmp.timer; enabled; vendor preset: enabled)
     Active: active (waiting) since Wed 2023-05-24 19:05:49 UTC; 3min 57s ago
    Trigger: Thu 2023-05-25 00:00:00 UTC; 4h 50min left
   Triggers: ● securedrop-clean-tmp.service
vagrant@app-staging:~$ systemctl status securedrop-submissions-today.service
● securedrop-submissions-today.service - job to clean SecureDrop tmp dir daily
     Loaded: loaded (/lib/systemd/system/securedrop-submissions-today.service; static; vendor preset: e>
     Active: inactive (dead)
TriggeredBy: ● securedrop-submissions-today.timer
vagrant@app-staging:~$ systemctl status securedrop-submissions-today.timer
● securedrop-submissions-today.timer - Run submissions-today 1h before daily reboot
     Loaded: loaded (/lib/systemd/system/securedrop-submissions-today.timer; enabled; vendor preset: en>
     Active: active (waiting) since Wed 2023-05-24 19:05:49 UTC; 4min 10s ago
    Trigger: Thu 2023-05-25 17:51:57 UTC; 22h left
   Triggers: ● securedrop-submissions-today.service
```

* [x] Wait a day
* [x] verify you receive the next "Submissions in the past 24h" email at the same time as before.

(Not configured on this staging server, but I've also installed this package on my QA instance, where it is.)

* [x] Run `sudo -u www-data ls -l /var/lib/securedrop/tmp/` and verify the `delete-me` file is gone.
* [x] Run `sudo cat /var/log/auth.log | grep cron | grep www-data`, verify the lines like `app CRON[268825]: pam_unix(cron:session): session opened for user www-data by (uid=0)` have stopped.

**Pull request merged**